#English

## 1. Introduction



In this project, an Exploratory Data Analysis (EDA) and a time series-based sales prediction model were performed using a dataset available at this [link](https://www.kaggle.com/datasets/ahmedabbas757/coffee-sales) from Kaggle.

The primary objective of this project is to understand historical sales patterns and develop a machine learning model capable of anticipating future behavior to provide valuable insights for decision-making.

The workflow included the following key stages:

1. **Data Extraction from Kaggle using an API**:
    The Kagglehub API was used to extract the data correctly and ensure it remains updated if there are changes to the data source.

2. **Exploratory Data Analysis (EDA)**:
    Python and libraries such as pandas, matplotlib, and seaborn were utilized to clean and analyze the dataset, identify trends, seasonality, and key variables affecting sales, as well as to quickly visualize the data.

3. **Predictive Modeling with Machine Learning**:
    The Prophet forecasting model developed by Facebook was implemented to predict future sales, taking into account trends and seasonal events.

4. **Visualization in Power BI**:
    The data and results obtained were integrated into an interactive Power BI dashboard, enabling the communication of insights in a visual and intuitive manner.

This project aims to demonstrate technical skills in data analysis, predictive modeling, and the creation of dynamic visualizations, as well as the ability to work with modern data science tools and business analytics. This approach is essential for extracting value from data and supporting evidence-based decision-making.

## Project Development

This section will detail step by step the process followed to develop the sales analysis and prediction, from setting up the environment and preparing the data to implementing the model and the final visualization.

### Importing Libraries

The first step was to import the necessary libraries to carry out the analysis and modeling. Standard Python tools widely recognized in the data science community were used, along with data visualization tools, the Kagglehub library for data extraction, and Prophet for model development.

In [3]:
# Libraries for data manipulation, handling, and analysis
import pandas as pd
import numpy as np

# Library for data extraction from Kaggle
import kagglehub
import os

# Libraries for visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Libraries for the time series model
from prophet import Prophet
from prophet.diagnostics import performance_metrics

# Handling warnings
import warnings
warnings.filterwarnings("ignore")

# Configuring the style of visualizations
plt.style.use('seaborn-v0_8-dark')

### Data Extraction

To ensure the analysis is based on up-to-date information, the Kagglehub library was used to download the data directly from Kaggle. This approach enables a dynamic and flexible workflow, as the information can be easily synchronized if updates to the dataset occur.

In [4]:
# The dataset is downloaded from Kaggle
path = kagglehub.dataset_download("ahmedabbas757/coffee-sales")

# Name of the downloaded file
file_name = "Coffee Shop Sales.xlsx"

# The full file path is constructed
file_path = os.path.join(path, file_name)

# The file is loaded into a pandas DataFrame
coffee_sales = pd.read_excel(file_path)

100%|██████████| 8.23M/8.23M [00:00<00:00, 23.8MB/s]

Extracting files...


### Exploratory Data Analysis (EDA)

Once the data is loaded into a DataFrame, it is essential to perform an initial exploration to understand its structure and data quality. Using the info() method, we get an overview of the data, showing key information about its structure.

In [5]:
coffee_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149116 entries, 0 to 149115
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   transaction_id    149116 non-null  int64         
 1   transaction_date  149116 non-null  datetime64[ns]
 2   transaction_time  149116 non-null  object        
 3   transaction_qty   149116 non-null  int64         
 4   store_id          149116 non-null  int64         
 5   store_location    149116 non-null  object        
 6   product_id        149116 non-null  int64         
 7   unit_price        149116 non-null  float64       
 8   product_category  149116 non-null  object        
 9   product_type      149116 non-null  object        
 10  product_detail    149116 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(4), object(5)
memory usage: 12.5+ MB


We can see that the data is organized by a transaction identifier, accompanied by important details such as the transaction date, quantity sold, store location, and product features, such as unit price, category, and type. Additionally, there are a total of 149,116 rows, and most importantly, none of them contain missing data. An additional point to note is that the date and time of the transaction are separated into different columns.

### Transforming the Date and Time Columns

Since the transaction date and time are in separate columns (`transaction_date` and `transaction_time`), the goal of this step is to combine them into a single column with a timestamp format (date and time) to facilitate temporal analysis.

In [6]:
# The 'transaction_date' and 'transaction_time' columns are concatenated into a single column called 'transaction_timestamp'
# First, both columns are converted to string (str) type and joined with a space in between
coffee_sales["transaction_timestamp"] = coffee_sales["transaction_date"].astype(str) + ' ' + coffee_sales["transaction_time"].astype(str)

# The new 'transaction_timestamp' column is converted to datetime type for easier temporal analysis
coffee_sales["transaction_timestamp"] = pd.to_datetime(coffee_sales["transaction_timestamp"])

# The original 'transaction_date' and 'transaction_time' columns are dropped
coffee_sales.drop(["transaction_date", "transaction_time"], axis=1, inplace=True)

To verify that the new column is correct, we can check the data by displaying the first few rows and ensuring that the transaction_timestamp column contains the expected datetime format.

In [7]:
coffee_sales["transaction_timestamp"].head()

0   2023-01-01 07:06:11
1   2023-01-01 07:08:56
2   2023-01-01 07:14:04
3   2023-01-01 07:20:24
4   2023-01-01 07:22:41
Name: transaction_timestamp, dtype: datetime64[ns]

#Español

##1. Introducción


En este proyecto, se realizó un análisis exploratorio de datos (EDA) y un modelo de predicción de ventas basado en series temporales usando un conjunto de datos disponible en este [enlace](https://www.kaggle.com/datasets/ahmedabbas757/coffee-sales) de Kaggle.

El Objetivo principal de este proyecto es comprender los patrones históricos de ventas y poder desarrollar un modelo de aprendizaje automático que pueda anticipar el comportamiento futuro para poder proporcionar información valiosa para para la toma de decisiones.

El flujo de trabajo incluyó las siguientes etapas principales:


1.   **Extracción de los datos de Kaggle usando una Api**:
    Se uso la API Kagglehub para poder extraer los datos de manera correcta y poder mantenerlos actualizados si existe un cambio en el origen de los datos.
2.   **Análisis Exploratorio de Datos (EDA)**:
    Se utilizó Python y bibliotecas como pandas, matplotlib y seaborn para limpiar y analizar el conjunto de datos, identificando tendencias, estacionalidad y variables clave que afectan las ventas, así como para visualizar los datos de manera rápida.
3. **Modelado Predictivo de aprendizaje automatico**:
    Se implementó el modelo de pronóstico Prophet desarrollado por Facebook para predecir ventas futuras, considerando aspectos como tendencias y eventos estacionales.
4. **Visualización en Power BI**:
    Los datos y resultados obtenidos se integraron en un tablero interactivo en Power BI, facilitando la comunicación de insights de manera visual e intuitiva.

El proyecto busca demostrar habilidades técnicas en análisis de datos, modelado predictivo y creación de visualizaciones dinámicas, así como la capacidad de trabajar con herramientas modernas de ciencia de datos y análisis empresarial. Este enfoque es crucial para extraer valor de los datos y apoyar la toma de decisiones basadas en evidencia.



## Desarrollo del proyecto

En esta sección, se detallará paso a paso el proceso seguido para desarrollar el análisis y la predicción de las ventas, desde la configuración del entorno y la preparación de los datos, hasta la implementación del modelo y la visualización final.

### Importación de las librerías

El primer paso fue importar las librerías necesarias para llevar a cabo el análisis y modelado. Se utilizaron herramientas estándar de Python que son ampliamente reconocidas en la comunidad de ciencia de datos, así como herramientas de visualización de datos, la libería de Kagglehub para la extracción de los datos y Prophet, para el desarrollo del modelo.

In [8]:
# Librerías para la manipulación, manejo  y análisis de datos
import pandas as pd
import numpy as np

# Librería para extracción de datos desde Kaggle
import kagglehub
import os

# Librerías para visualización
import matplotlib.pyplot as plt
import seaborn as sns

# Librerías para el modelo basado en series temporales
from prophet import Prophet
from prophet.diagnostics import performance_metrics

# Manejo de advertencias
import warnings
warnings.filterwarnings("ignore")

# Configuración del estilo de las visualizaciones
plt.style.use('seaborn-v0_8-dark')


### Extracción de los Datos

Para garantizar que el análisis se base en información al día, se utilizó la librería kagglehub para descargar los datos directamente desde Kaggle, esto permite tener un enfoque en un flujo de trabajo dinámico y flexible, ya que la información se puede sincronizar fácilmente en caso de existir actualizaciones en el conjunto de datos.

In [9]:
# Se descarga el conjunto de datos desde Kaggle
path = kagglehub.dataset_download("ahmedabbas757/coffee-sales")

# Nombre del archivo descargado
file_name = "Coffee Shop Sales.xlsx"

# Se construye la ruta completa del archivo
file_path = os.path.join(path, file_name)

# Se carga el archivo a un DataFrame de pandas.
coffee_sales = pd.read_excel(file_path)

### Análisis Exploratorio de Datos (EDA)

Una vez cargados los datos en un DataFrame, es fundamental realizar una exploración inicial para comprender su estructura y calidad de los datos usando el método info(), este nos proporciona una vista general de los datos para ver información clave sobre la estructura de los datos.

In [10]:
coffee_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149116 entries, 0 to 149115
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   transaction_id    149116 non-null  int64         
 1   transaction_date  149116 non-null  datetime64[ns]
 2   transaction_time  149116 non-null  object        
 3   transaction_qty   149116 non-null  int64         
 4   store_id          149116 non-null  int64         
 5   store_location    149116 non-null  object        
 6   product_id        149116 non-null  int64         
 7   unit_price        149116 non-null  float64       
 8   product_category  149116 non-null  object        
 9   product_type      149116 non-null  object        
 10  product_detail    149116 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(4), object(5)
memory usage: 12.5+ MB


Podemos ver que los datos están organizados por un identificador de transacción, el cual está acompañado de información importante como la fecha de la transacción, la cantidad vendida, la ubicación de la tienda y características del producto, como su precio unitario, categoría y tipo.
Además de estos detalles, se cuenta con un total de 149,116 filas, y lo más relevante es que ninguna de ellas contiene datos nulos. Un aspecto adicional a destacar es que la fecha y la hora de la transacción están separadas en columnas distintas.



### Transformación de las columnas de fecha y hora

Dado que la fecha y la hora de la transacción se encuentran separadas en columnas distintas (`transaction_date` y `transaction_time`), el objetivo de este paso es combinarlas en una sola columna con un formato de timestamp (fecha y hora) para facilitar el análisis temporal.

In [11]:
# Se concadenan las columnas 'transaction_date' y 'transaction_time' en una sola columna llamada 'transaction_timestamp'
# Primero, se convierte ambas columnas a tipo cadena (str) y se unen con un espacio en medio
coffee_sales["transaction_timestamp"] = coffee_sales["transaction_date"].astype(str) + ' ' + coffee_sales["transaction_time"].astype(str)

# Se convierte la columna nueva 'transaction_timestamp' a tipo datetime para facilitar el análisis temporal
coffee_sales["transaction_timestamp"] = pd.to_datetime(coffee_sales["transaction_timestamp"])

# Se eliminan las columnas originales 'transaction_date' y 'transaction_time'
coffee_sales.drop(["transaction_date", "transaction_time"], axis=1, inplace=True)

Para verificar que la nueva columna es correcta, podemos comprobar los datos mostrando las primeras filas y asegurándonos de que la columna transaction_timestamp contiene el formato datetime esperado.


In [12]:
coffee_sales["transaction_timestamp"].head()

0   2023-01-01 07:06:11
1   2023-01-01 07:08:56
2   2023-01-01 07:14:04
3   2023-01-01 07:20:24
4   2023-01-01 07:22:41
Name: transaction_timestamp, dtype: datetime64[ns]

In [13]:
coffee_sales["transaction_id"].drop_duplicates().shape

(149116,)

In [14]:
coffee_sales[["store_location", "store_id"]].value_counts()

store_location   store_id
Hell's Kitchen   8           50735
Astoria          3           50599
Lower Manhattan  5           47782
Name: count, dtype: int64

In [15]:
coffee_sales["product_detail"].drop_duplicates().value_counts()

product_detail
Ethiopia Rg                  1
Spicy Eye Opener Chai Lg     1
Chocolate syrup              1
Sugar Free Vanilla syrup     1
Organic Decaf Blend          1
                            ..
Columbian Medium Roast Rg    1
Brazilian Lg                 1
Brazilian Rg                 1
Latte                        1
I Need My Bean! Diner mug    1
Name: count, Length: 80, dtype: int64

In [16]:
coffee_sales["unit_price"].max()

45.0

In [17]:
coffee_sales[coffee_sales["unit_price"]==45]

transaction_id  transaction_qty  store_id   store_location  \
5182              5212                1         5  Lower Manhattan   
5874              5904                1         8   Hell's Kitchen   
5939              5969                1         5  Lower Manhattan   
7616              7646                1         8   Hell's Kitchen   
8443              8473                1         3          Astoria   
...                ...              ...       ...              ...   
144743          145080                1         5  Lower Manhattan   
145411          145748                1         5  Lower Manhattan   
148573          148914                1         8   Hell's Kitchen   
148702          149043                8         8   Hell's Kitchen   
148784          149125                1         5  Lower Manhattan   

        product_id  unit_price product_category   product_type product_detail  \
5182             8        45.0     Coffee beans  Premium Beans      Civet Cat   
5874             8        45.0     Coffee beans  Premium Beans      Civet Cat   
5939             8        45.0     Coffee beans  Premium Beans      Civet Cat   
7616             8        45.0     Coffee beans  Premium Beans      Civet Cat   
8443             8        45.0     Coffee beans  Premium Beans      Civet Cat   
...            ...         ...              ...            ...            ...   
144743           8        45.0     Coffee beans  Premium Beans      Civet Cat   
145411           8        45.0     Coffee beans  Premium Beans      Civet Cat   
148573           8        45.0     Coffee beans  Premium Beans      Civet Cat   
148702           8        45.0     Coffee beans  Premium Beans      Civet Cat   
148784           8        45.0     Coffee beans  Premium Beans      Civet Cat   

       transaction_timestamp  
5182     2023-01-10 09:20:43  
5874     2023-01-11 11:24:31  
5939     2023-01-11 13:53:09  
7616     2023-01-14 10:57:38  
8443     2023-01-15 18:18:51  
...                      ...  
144743   2023-06-27 07:49:34  
145411   2023-06-27 12:12:41  
148573   2023-06-30 10:25:26  
148702   2023-06-30 11:18:31  
148784   2023-06-30 13:03:56  

[190 rows x 10 columns]

In [18]:
coffee_sales["unit_price"].min()

0.8

In [19]:
coffee_sales[coffee_sales["unit_price"]==0.8]

transaction_id  transaction_qty  store_id   store_location  \
3254              3284                2         5  Lower Manhattan   
3291              3321                1         5  Lower Manhattan   
3318              3348                1         8   Hell's Kitchen   
3325              3355                1         5  Lower Manhattan   
3331              3361                1         5  Lower Manhattan   
...                ...              ...       ...              ...   
148994          149335                1         8   Hell's Kitchen   
149013          149354                2         5  Lower Manhattan   
149031          149372                1         3          Astoria   
149059          149400                2         3          Astoria   
149115          149456                2         8   Hell's Kitchen   

        product_id  unit_price product_category      product_type  \
3254            64         0.8         Flavours     Regular syrup   
3291            63         0.8         Flavours     Regular syrup   
3318            65         0.8         Flavours  Sugar free syrup   
3325            65         0.8         Flavours  Sugar free syrup   
3331            64         0.8         Flavours     Regular syrup   
...            ...         ...              ...               ...   
148994          63         0.8         Flavours     Regular syrup   
149013          63         0.8         Flavours     Regular syrup   
149031          84         0.8         Flavours     Regular syrup   
149059          65         0.8         Flavours  Sugar free syrup   
149115          64         0.8         Flavours     Regular syrup   

                  product_detail transaction_timestamp  
3254              Hazelnut syrup   2023-01-07 06:27:23  
3291                Carmel syrup   2023-01-07 07:44:17  
3318    Sugar Free Vanilla syrup   2023-01-07 08:09:37  
3325    Sugar Free Vanilla syrup   2023-01-07 08:18:07  
3331              Hazelnut syrup   2023-01-07 08:20:30  
...                          ...                   ...  
148994              Carmel syrup   2023-06-30 17:33:45  
149013              Carmel syrup   2023-06-30 17:59:31  
149031           Chocolate syrup   2023-06-30 18:11:40  
149059  Sugar Free Vanilla syrup   2023-06-30 18:44:37  
149115            Hazelnut syrup   2023-06-30 20:57:19  

[6790 rows x 10 columns]

In [20]:
coffee_sales["product_category"].value_counts()

product_category
Coffee                58416
Tea                   45449
Bakery                22796
Drinking Chocolate    11468
Flavours               6790
Coffee beans           1753
Loose Tea              1210
Branded                 747
Packaged Chocolate      487
Name: count, dtype: int64

In [21]:
coffee_sales["product_type"].value_counts()

product_type
Brewed Chai tea          17183
Gourmet brewed coffee    16912
Barista Espresso         16403
Hot chocolate            11468
Brewed Black tea         11350
Brewed herbal tea        11245
Scone                    10173
Organic brewed coffee     8489
Drip coffee               8477
Premium brewed coffee     8135
Pastry                    6912
Biscotti                  5711
Brewed Green tea          5671
Regular syrup             4979
Sugar free syrup          1811
Housewares                 526
Chai tea                   443
Organic Beans              415
Gourmet Beans              366
Premium Beans              336
Espresso Beans             319
Herbal tea                 305
Black tea                  303
Drinking Chocolate         266
Organic Chocolate          221
Clothing                   221
House blend Beans          183
Green tea                  159
Green beans                134
Name: count, dtype: int64

In [22]:
coffee_sales["product_detail"].value_counts().shape

(80,)

In [23]:
coffee_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149116 entries, 0 to 149115
Data columns (total 10 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   transaction_id         149116 non-null  int64         
 1   transaction_qty        149116 non-null  int64         
 2   store_id               149116 non-null  int64         
 3   store_location         149116 non-null  object        
 4   product_id             149116 non-null  int64         
 5   unit_price             149116 non-null  float64       
 6   product_category       149116 non-null  object        
 7   product_type           149116 non-null  object        
 8   product_detail         149116 non-null  object        
 9   transaction_timestamp  149116 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(4), object(4)
memory usage: 11.4+ MB


In [24]:
coffee_sales["transaction_id"] = coffee_sales["transaction_id"].astype("category")
coffee_sales["store_id"] = coffee_sales["store_id"].astype("category")
coffee_sales["store_location"] = coffee_sales["store_location"].astype("category")
coffee_sales["product_id"] = coffee_sales["product_id"].astype("category")
coffee_sales["product_category"] = coffee_sales["product_category"].astype("category")
coffee_sales["product_type"] = coffee_sales["product_type"].astype("category")
coffee_sales["product_detail"] = coffee_sales["product_detail"].astype("category")
coffee_sales["transaction_qty"] = coffee_sales["transaction_qty"].astype("int16")
coffee_sales["unit_price"] = coffee_sales["unit_price"].astype("float32")